apiserver/facades/client/modelgeneration/modelgeneration.go

```diff
@@ -148,19 +150,54 @@ func (api *API) CommitBranch(arg params.BranchArg) (params.IntResult, error) {
 		return result, common.ErrPerm
 	}
 
-	generation, err := api.model.Branch(arg.BranchName)
+	branch, err := api.model.Branch(arg.BranchName)
 	if err != nil {
-		return result, errors.Trace(err)
+		return intResultsError(err)
+	}
+
+	if err = api.applyBranchConfig(branch); err != nil {
+		return intResultsError(err)
 	}
 
-	if genId, err := generation.Commit(api.apiUser.Name()); err != nil {
+	if genId, err := branch.Commit(api.apiUser.Name()); err != nil {
 		result.Error = common.ServerError(err)
 	} else {
 		result.Result = genId
 	}
 	return result, nil
 }
 
+// applyBranchConfig applies configuration deltas from a branch to the
+// appropriate application charm settings documents and writes them back
+// to the DB. This writes the ultimate data as a single transaction,
+// so a branch is either committed without error, or there are no changes.
+func (api *API) applyBranchConfig(branch Generation) error {
+	var allSettings []state.Settings
+
+	// Collect up all the new application settings collections that will result
+	// from committing the branch.
+	for appName, delta := range branch.Config() {
+		if len(delta) == 0 {
+			continue
+		}
+		app, err := api.st.Application(appName)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		cfg, err := app.CharmSettingsWithDelta(delta)
```

Comment:
My concern with this is that reading the CharmSettings should be part of the transaction, with some sort of assert that they haven't changed before we write them out again.
Otherwise concurrent updates will step on eachothers toes in ways we may not expect. (Namely, a branch that changes foo, but negate a change to 'bar' because we read the config before bar was update, and wrote it out.)
I believe the Settings code actually makes sure to only "$set: {doc.SETTING: newval}" so it doesn't touch the other values.
But that would need to be talking in terms of "deltas", not in terms of "allSettings".

Am I missing something?